Redo cross-referencing

mrerrormessage opened the pull request:
@nicolaspayette , @bainco , in order to simplify and speed the build process, I'm changing the way NetLogo performs model cross-referencing. I've done this by adding a file `crossReference.conf` in the root of the models directory which defines all of the paths which are referenced. These virtual paths will be displayed in the Models Library dialog but will not be created on the file system at any point. This should let us remove a massive section of the gitignore (though I haven't done that yet) and avoid copying files all over the place when building the models library. Additionally, adding cross references to the models library no longer requires changes to the NetLogo project. Finally, we no longer have to copy around auxiliary (non-.nlogo/.nlogo3d) files for model preview images and models which require external files.

Things I would like feedback on:

1. Does the format of `crossReference.conf` make sense? It's possible (and fairly easy) to change it now. This could mean changing names, conventions, or even the shape of the file. The syntax used is the fairly-flexible [HOCON](https://github.com/typesafehub/config/blob/master/HOCON.md) which is sort of like JSON + java properties, but writable by human beings.
2. Would you prefer that `crossReference.conf` live somewhere else? It *could* live in the NetLogo project or in a subdirectory of the models project (NetLogo could copy it to the root when it needs it).
3. Would you like documentation about `crossReference.conf` added anywhere in the models project and/or a wiki? 

The other (minor) change here is that I've made the index.txt file an ".conf" file as well. It's still generated by NetLogo.

nicolaspayette commented:
@mrerrormessage, I have just rediscovered the relevant issue in the main repo: https://github.com/NetLogo/NetLogo/issues/933. I still think the solution I proposed over there has merits (and I'm happy to expand on it if you would like me too), but the `.conf` file solution is already a big improvement on the previous situation.

Question: does it allow to duplicate one model in multiple locations? I don't think we have any of those at the moment, and I don't think it would be wise to add any, but we never know.

> Does the format of crossReference.conf make sense?

Yes.

> Would you prefer that crossReference.conf live somewhere else?

No.

> Would you like documentation about crossReference.conf added anywhere in the models project and/or a wiki?

Yes. https://github.com/NetLogo/NetLogo/wiki/Preparing-the-Models-Library-for-release#updating-the-multiply-categorized-models-in-binreleasesh would need to be updated. It could refer to a new, separate page that lives in models wiki instead of the main wiki.

Another thing I would like to see is a models repo automated test that checks that all `source` and `referenceIn` paths in the `.conf` file are valid. This way, if we move a model or a folder around and forget to update the `.conf` file, we don't need to wait for the main NetLogo build to break before finding out.

That being said, I still think that it would be worth exploring the possibility of getting rid of `index.txt` (or `index.conf`) entirely. My intuition is that the cost of building the model index in some background thread when NetLogo starts would be pretty low. There is some information duplication at the moment: all the info in `index.txt` is already contained in the file system. It's also annoying to have to rebuild the index from the main NetLogo build when working on models: it would be nice to just drop a model in its destination folder and have it appear in the library when restarting NetLogo.

mrerrormessage commented:
Thanks for the feedback. I wanted to go through and discuss a couple of items you mentioned:

> I have just rediscovered the relevant issue in the main repo: NetLogo/NetLogo#933

I think that this idea has some potential. This change captures most of the benefits that 933 mentions, although it does use a different approach. Since the ModelsLibrary (and it's collaborator ModelsLibraryDialog) have been converted to scala, the change would make it easier to take future steps in that direction. I also think there are some benefits to this that the 933 approach wouldn't have. For instance, in this approach, some of the `referenceIn` directories don't actually exist on the file system, they're created only for dialog display purposes. More on that later.

> Yes. https://github.com/NetLogo/NetLogo/wiki/Preparing-the-Models-Library-for-release#updating-the-multiply-categorized-models-in-binreleasesh would need to be updated. It could refer to a new, separate page that lives in models wiki instead of the main wiki.

I'll create a page on the models wiki and link it. We don't use `bin/release.sh` anymore, so I can't imagine that the NetLogo wiki page is fully up-to-date. We're also in the process of renovating our wiki and the fewer changes we make on it now the easier the renovation will be.

> Another thing I would like to see is a models repo automated test that checks that all source and referenceIn paths in the .conf file are valid.

I will add a test that checks that all of the `source` paths are valid (for both individual models and directories. The `referenceIn` paths don't actually need to be paths on the file system. The Models Library will create folders referencing them even if there is no folder at that path on the file system. Three folders which don't exist on the filesystem which have entries thanks to cross referencing are "Code Examples/Extension Examples/arduino", "Code Examples/Extension Examples/bitmap", and "Curricular Models/EACH/Unverified". I think this represents the big advantage over the symlink approach: directories are only created to put actual models in and no directories are created for cross-referencing purposes only. 

> That being said, I still think that it would be worth exploring the possibility of getting rid of index.txt (or index.conf) entirely. ...

I'm not definitely opposed to having NetLogo re-index the models library on startup. I think there are a number of reasons why this would be a good idea, not least that it makes models library maintenance easier. I'm hesitant to totally remove an index file from the distribution for several reasons:

1. File system access can be slow, especially on older computers. Building the index in the background sounds great in theory, but we want the models library to work for the built-in models immediately. Building indexes for 400+ models could take several seconds and we don't want to give users the impression that the models library (or NetLogo in general) is "slow" by displaying a waiting dialog while we build the index.
2. We don't always have write access to the models library directory once NetLogo is installed. This makes it difficult to cache a runtime-generated index between runs, meaning we would have to store an updated cache in-memory each time we start NetLogo.

Thanks again for the feedback! I would like to discuss adding runtime-generated indexing at some point after 6.0.2 is released, potentially for the release following that (depending on other items which are on the CCL development docket).

mrerrormessage commented:
OK, I've updated this PR and added a wiki page documenting [here](https://github.com/NetLogo/models/wiki/Models-Cross-Referencing). I will add a reference to this new page in the NetLogo wiki.

bainco commented on crossReference.conf:
```diff
@@ -0,0 +1,55 @@
+
+  // IABM Textbook models duplicated in Sample Models and Code Examples
+  { source: "IABM Textbook/chapter 3/El Farol Extensions/El Farol.nlogo",  referenceIn: "Sample Models/Social Science" }
+  { source: "IABM Textbook/chapter 3/DLA extensions/DLA Simple.nlogo",     referenceIn: "Sample Models/Chemistry & Physics/Diffusion Limited Aggregation" }
```
This line is producing an error. Should it be (capitalize "extensions"):

`{ source: "IABM Textbook/chapter 3/DLA Extensions/DLA Simple.nlogo",     referenceIn: "Sample Models/Chemistry & Physics/Diffusion Limited Aggregation" }`

I assume that means this new method is case-sensitive while the old was not?

mrerrormessage replied:
Now updated.
